File: 198-house-robber/house-robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
    // bottom-up optimized space complexity
    if(nums.length == 1) return nums[0]
    let backOne =  Math.max(nums[0],nums[1]);
    let backTwo = nums[0];

    for(let i=2;i<nums.length;i++){
        let temp = backOne;
        backOne = Math.max(backOne,backTwo + nums[i]);
        backTwo = temp;
    }
    return backOne;

};

File: 198-house-robber/house-robber.java
class Solution {
    public int rob(int[] nums) {

        if (nums.length == 1)
            return nums[0];

        int length = nums.length;
        int[] max = new int[length];

        max[0] = nums[0];
        max[1] = Math.max(nums[0], nums[1]);
        int ans = max[1];

        for (int i = 2; i < length; i++) {
            // rob the current / no rob
            ans = Math.max(Math.max(nums[i] + max[i - 2], nums[i - 1]), ans);
            max[i] = ans;

        }

        return ans;
    }
}